File: sdk/runanywhere-swift/Sources/RunAnywhere/Features/TTS/Analytics/TTSEvent.swift
//
//  TTSEvent.swift
//  RunAnywhere SDK
//
//  All TTS-related events in one place.
//  Each event declares its destination (public, analytics, or both).
//
//  Note: TTSEvent conforms to TelemetryEventProperties for strongly typed analytics.
//  This avoids string conversion/parsing and enables compile-time type checking.
//

import CRACommons
import Foundation

// MARK: - TTS Event

/// All TTS (Text-to-Speech) related events.
///
/// Usage:
/// ```swift
/// EventPublisher.shared.track(TTSEvent.synthesisCompleted(...))
/// ```
///
/// TTSEvent provides strongly typed properties via `telemetryProperties`.
/// This enables:
/// - Type safety at compile time
/// - No string parsing for analytics
/// - Validation guardrails
public enum TTSEvent: SDKEvent, TelemetryEventProperties {

    // MARK: - Model Lifecycle

    case modelLoadStarted(modelId: String, modelSizeBytes: Int64 = 0, framework: InferenceFramework = .unknown)
    case modelLoadCompleted(modelId: String, durationMs: Double, modelSizeBytes: Int64 = 0, framework: InferenceFramework = .unknown)
    case modelLoadFailed(modelId: String, error: SDKError, framework: InferenceFramework = .unknown)
    case modelUnloaded(modelId: String)

    // MARK: - Synthesis

    /// Synthesis started event
    /// - Parameters:
    ///   - characterCount: Number of characters in the text to synthesize
    ///   - sampleRate: Audio sample rate in Hz (default: RAC_TTS_DEFAULT_SAMPLE_RATE)
    case synthesisStarted(
        synthesisId: String,
        modelId: String,
        characterCount: Int,
        sampleRate: Int = Int(RAC_TTS_DEFAULT_SAMPLE_RATE),
        framework: InferenceFramework = .unknown
    )

    /// Streaming synthesis chunk generated
    case synthesisChunk(synthesisId: String, chunkSize: Int)

    /// Synthesis completed event
    /// - Parameters:
    ///   - audioDurationMs: Duration of generated audio in milliseconds
    ///   - audioSizeBytes: Size of generated audio in bytes
    ///   - processingDurationMs: Time taken to synthesize (processing time)
    ///   - charactersPerSecond: Synthesis speed (characters processed per second)
    ///   - sampleRate: Audio sample rate in Hz
    case synthesisCompleted(
        synthesisId: String,
        modelId: String,
        characterCount: Int,
        audioDurationMs: Double,
        audioSizeBytes: Int,
        processingDurationMs: Double,
        charactersPerSecond: Double,
        sampleRate: Int = Int(RAC_TTS_DEFAULT_SAMPLE_RATE),
        framework: InferenceFramework = .unknown
    )
    case synthesisFailed(synthesisId: String, modelId: String, error: SDKError)

    // MARK: - SDKEvent Conformance

    public var type: String {
        switch self {
        case .modelLoadStarted: return "tts_model_load_started"
        case .modelLoadCompleted: return "tts_model_load_completed"
        case .modelLoadFailed: return "tts_model_load_failed"
        case .modelUnloaded: return "tts_model_unloaded"
        case .synthesisStarted: return "tts_synthesis_started"
        case .synthesisChunk: return "tts_synthesis_chunk"
        case .synthesisCompleted: return "tts_synthesis_completed"
        case .synthesisFailed: return "tts_synthesis_failed"
        }
    }

    public var category: EventCategory { .tts }

    public var destination: EventDestination {
        switch self {
        case .synthesisChunk:
            // Chunk events are too chatty for public API
            return .analyticsOnly
        default:
            return .all
        }
    }

    public var properties: [String: String] {
        // Use derived properties from telemetryProperties for consistency
        // This eliminates duplicate property conversion logic
        telemetryProperties.toDictionary()
    }

    // MARK: - TelemetryEventProperties Conformance

    /// Strongly typed telemetry properties - no string conversion needed.
    /// These values are used directly by TelemetryEventPayload.
    public var telemetryProperties: TelemetryProperties {
        switch self {
        case .modelLoadStarted(let modelId, let modelSizeBytes, let framework):
            return TelemetryProperties(
                modelId: modelId,
                framework: framework.rawValue,
                voice: modelId,
                modelSizeBytes: modelSizeBytes > 0 ? modelSizeBytes : nil
            )

        case .modelLoadCompleted(let modelId, let durationMs, let modelSizeBytes, let framework):
            return TelemetryProperties(
                modelId: modelId,
                framework: framework.rawValue,
                processingTimeMs: durationMs,
                success: true,
                voice: modelId,
                modelSizeBytes: modelSizeBytes > 0 ? modelSizeBytes : nil
            )

        case .modelLoadFailed(let modelId, let error, let framework):
            return TelemetryProperties(
                modelId: modelId,
                framework: framework.rawValue,
                success: false,
                errorMessage: error.message,
                errorCode: error.code.rawValue,
                voice: modelId
            )

        case .modelUnloaded(let modelId):
            return TelemetryProperties(
                modelId: modelId,
                voice: modelId
            )

        case .synthesisStarted(let id, let modelId, let characterCount, let sampleRate, let framework):
            return TelemetryProperties(
                modelId: modelId,
                framework: framework.rawValue,
                characterCount: characterCount,
                sampleRate: sampleRate,
                voice: modelId,
                synthesisId: id
            )

        case .synthesisChunk(let id, let chunkSize):
            return TelemetryProperties(
                audioSizeBytes: chunkSize,
                synthesisId: id
            )

        case .synthesisCompleted(
            let id,
            let modelId,
            let charCount,
            let audioDurationMs,
            let audioSize,
            let processingDurationMs,
            let charsPerSecond,
            let sampleRate,
            let framework
        ):
            return TelemetryProperties(
                modelId: modelId,
                framework: framework.rawValue,
                processingTimeMs: processingDurationMs,
                success: true,
                characterCount: charCount,
                charactersPerSecond: charsPerSecond,
                audioSizeBytes: audioSize,
                sampleRate: sampleRate,
                voice: modelId,
                outputDurationMs: audioDurationMs,
                synthesisId: id
            )

        case .synthesisFailed(let id, let modelId, let error):
            return TelemetryProperties(
                modelId: modelId,
                success: false,
                errorMessage: error.message,
                errorCode: error.code.rawValue,
                voice: modelId,
                synthesisId: id
            )
        }
    }
}

File: sdk/runanywhere-flutter/packages/runanywhere/lib/features/tts/analytics/tts_event.dart
import 'package:runanywhere/public/events/sdk_event.dart';

/// All TTS (Text-to-Speech) related events.
///
/// Usage:
/// ```dart
/// EventPublisher.shared.track(TTSSynthesisCompletedEvent(...));
/// ```
///
/// Matches iOS `TTSEvent` enum from TTSEvent.swift
sealed class TTSEvent with SDKEventDefaults {
  const TTSEvent();

  @override
  EventCategory get category => EventCategory.tts;
}

// MARK: - Model Lifecycle Events

/// Model load started
class TTSModelLoadStartedEvent extends TTSEvent {
  final String voiceId;
  final int modelSizeBytes;
  final String framework;

  const TTSModelLoadStartedEvent({
    required this.voiceId,
    this.modelSizeBytes = 0,
    this.framework = 'unknown',
  });

  @override
  String get type => 'tts_model_load_started';

  @override
  Map<String, String> get properties {
    final props = <String, String>{
      'voice_id': voiceId,
      'framework': framework,
    };
    if (modelSizeBytes > 0) {
      props['model_size_bytes'] = modelSizeBytes.toString();
    }
    return props;
  }
}

/// Model load completed
class TTSModelLoadCompletedEvent extends TTSEvent {
  final String voiceId;
  final double durationMs;
  final int modelSizeBytes;
  final String framework;

  const TTSModelLoadCompletedEvent({
    required this.voiceId,
    required this.durationMs,
    this.modelSizeBytes = 0,
    this.framework = 'unknown',
  });

  @override
  String get type => 'tts_model_load_completed';

  @override
  Map<String, String> get properties {
    final props = <String, String>{
      'voice_id': voiceId,
      'duration_ms': durationMs.toStringAsFixed(1),
      'framework': framework,
    };
    if (modelSizeBytes > 0) {
      props['model_size_bytes'] = modelSizeBytes.toString();
    }
    return props;
  }
}

/// Model load failed
class TTSModelLoadFailedEvent extends TTSEvent {
  final String voiceId;
  final String error;
  final String framework;

  const TTSModelLoadFailedEvent({
    required this.voiceId,
    required this.error,
    this.framework = 'unknown',
  });

  @override
  String get type => 'tts_model_load_failed';

  @override
  Map<String, String> get properties => {
        'voice_id': voiceId,
        'error': error,
        'framework': framework,
      };
}

/// Model unloaded
class TTSModelUnloadedEvent extends TTSEvent {
  final String voiceId;

  const TTSModelUnloadedEvent({required this.voiceId});

  @override
  String get type => 'tts_model_unloaded';

  @override
  Map<String, String> get properties => {'voice_id': voiceId};
}

// MARK: - Synthesis Events

/// Synthesis started
class TTSSynthesisStartedEvent extends TTSEvent {
  final String synthesisId;
  final String voiceId;
  final int characterCount;
  final String framework;

  const TTSSynthesisStartedEvent({
    required this.synthesisId,
    required this.voiceId,
    required this.characterCount,
    this.framework = 'unknown',
  });

  @override
  String get type => 'tts_synthesis_started';

  @override
  Map<String, String> get properties => {
        'synthesis_id': synthesisId,
        'voice_id': voiceId,
        'character_count': characterCount.toString(),
        'framework': framework,
      };
}

/// Synthesis chunk generated (streaming)
class TTSSynthesisChunkEvent extends TTSEvent {
  final String synthesisId;
  final int chunkSize;

  const TTSSynthesisChunkEvent({
    required this.synthesisId,
    required this.chunkSize,
  });

  @override
  String get type => 'tts_synthesis_chunk';

  /// Chunk events are too chatty for public API
  @override
  EventDestination get destination => EventDestination.analyticsOnly;

  @override
  Map<String, String> get properties => {
        'synthesis_id': synthesisId,
        'chunk_size': chunkSize.toString(),
      };
}

/// Synthesis completed
class TTSSynthesisCompletedEvent extends TTSEvent {
  final String synthesisId;
  final String voiceId;
  final int characterCount;
  final double audioDurationMs;
  final int audioSizeBytes;
  final double processingDurationMs;
  final double charactersPerSecond;
  final String framework;

  const TTSSynthesisCompletedEvent({
    required this.synthesisId,
    required this.voiceId,
    required this.characterCount,
    required this.audioDurationMs,
    required this.audioSizeBytes,
    required this.processingDurationMs,
    required this.charactersPerSecond,
    this.framework = 'unknown',
  });

  @override
  String get type => 'tts_synthesis_completed';

  @override
  Map<String, String> get properties => {
        'synthesis_id': synthesisId,
        'voice_id': voiceId,
        'character_count': characterCount.toString(),
        'audio_duration_ms': audioDurationMs.toStringAsFixed(1),
        'audio_size_bytes': audioSizeBytes.toString(),
        'processing_duration_ms': processingDurationMs.toStringAsFixed(1),
        'chars_per_second': charactersPerSecond.toStringAsFixed(2),
        'framework': framework,
      };
}

/// Synthesis failed
class TTSSynthesisFailedEvent extends TTSEvent {
  final String synthesisId;
  final String error;

  const TTSSynthesisFailedEvent({
    required this.synthesisId,
    required this.error,
  });

  @override
  String get type => 'tts_synthesis_failed';

  @override
  Map<String, String> get properties => {
        'synthesis_id': synthesisId,
        'error': error,
      };
}
